File: modules/static/js/filter.js
/*
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.
*/

'use strict';

/******************************************************************************
 *
 * Filter subsystem
 *
 */

/**
 * Get a filter ID by name. This is the primary API for functions preparing query
 * strings to make filtered requests with pre-defined/named filters.
 */
mc.filter = function(name)
{
	return maybe(() => mc.filter.filter[name].filter_id);
};

/**
 * This holds our predefined filters. They will obtain ID's when registered with
 * the server. A container in mc.session also holds ID's obtained with a string
 * as a value naming the filter in this object.
 */
mc.filter.filter =
{
	// Primary filter
	sync:
	{
		// ID given by server. We cross-reference in mc.session.filter[id]
		filter_id: undefined,

		// Actual filter
		content:
		{
		},
	},

	initial_sync:
	{
		content:
		{
			room:
			{
				not_rooms:
				[
					//"!vfFxDRtZSSdspfTSEr:matrix.org",
				],

				state:
				{
					not_types:
					[
						"m.room.member",
					],
				},
			},
		},
	},

	image_search:
	{
		content:
		{
			event_fields:
			[
//				"event_id",
				//"content.msgtype",
			],
/*
			room:
			{
				timeline:
				{
					types:
					[
						//"m.room.message",
					],
				},
			},
*/
		},
	},
};

Object.clean(mc.filter.filter);


/* Filter initialization on client load. Checks the filter ID's remembered
 * from storage and posts any new filters the server doesn't know about.
 *
 * TODO: This will be removed and the primary filter API will manage the filters as
 * it's called instead using await to post filters and return with IDs if necessary.
 */
mc.filter.init = async function(opts = {})
{
	for(let filter_id in mc.session.filter)
	{
		let name = mc.session.filter[filter_id];
		if(!(name in mc.filter.filter))
			continue;

		let item = mc.filter.filter[name];
		if(item.filter_id !== undefined)
			continue;

		let res = await mc.filter.init.get(filter_id);
		if(res === false)
			console.log("Purged unused filter id[" + filter_id + "]");
		else
			console.info("Associated filter '" + name + "' id[" + filter_id + "]");
	}

	for(let name in mc.filter.filter)
	{
		let item = mc.filter.filter[name];
		if(item.filter_id !== undefined)
			continue;

		let filter_id = await mc.filter.init.post(name);
		console.log("Posted filter '" + name + "' id[" + filter_id + "]");
	}
};

/** Attempt to get a filter we have an ID for and associate it
 * @returns the name for the filter ID. If no name, the filter
 * ID is forgotten to optimize the next init.
 */
mc.filter.init.get = async function(filter_id, opts = {})
{
	let request = mc.m.filter.get(filter_id, opts); try
	{
		let data = await request.response;
		let filter = JSON.stringify(data);
		for(let name in mc.filter.filter)
		{
			let item = mc.filter.filter[name];
			let content = JSON.stringify(item.content);
			if(filter == content)
			{
				item.filter_id = filter_id;
				mc.session.filter[filter_id] = name;
				return name;
			}
		}

		delete mc.session.filter[filter_id];
		return false;
	}
	catch(error)
	{
		switch(error.status)
		{
			case 404:
				delete mc.session.filter[filter_id];
				return false;

			default:
				throw error;
		}
	}
};

/** Posts a new filter, associates the new ID and saves it to session
 * @returns the filter_id for the name
 */
mc.filter.init.post = async function(name, opts = {})
{
	let item = mc.filter.filter[name];
	let request = mc.m.filter.post(item.content, opts);
	let data = await request.response;
	item.filter_id = data.filter_id;
	mc.session.filter[data.filter_id] = name;
	return item.filter_id;
};

File: ircd/m/filter.cc
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.

#include <ircd/m/m.h>

//TODO: globular expression
bool
ircd::m::match(const room_event_filter &filter,
               const event &event)
{
	for(const auto &room_id : json::get<"not_rooms"_>(filter))
		if(at<"room_id"_>(event) == unquote(room_id))
			return false;

	if(empty(json::get<"rooms"_>(filter)))
		return match(event_filter{filter}, event);

	for(const auto &room_id : json::get<"rooms"_>(filter))
		if(at<"room_id"_>(event) == unquote(room_id))
			return match(event_filter{filter}, event);

	return false;
}

//TODO: globular expression
bool
ircd::m::match(const event_filter &filter,
               const event &event)
{
	for(const auto &type : json::get<"not_types"_>(filter))
		if(at<"type"_>(event) == unquote(type))
			return false;

	for(const auto &sender : json::get<"not_senders"_>(filter))
		if(at<"sender"_>(event) == unquote(sender))
			return false;

	if(empty(json::get<"senders"_>(filter)) && empty(json::get<"types"_>(filter)))
		return true;

	if(empty(json::get<"senders"_>(filter)))
	{
		for(const auto &type : json::get<"types"_>(filter))
			if(at<"type"_>(event) == unquote(type))
				return true;

		return false;
	}

	if(empty(json::get<"types"_>(filter)))
	{
		for(const auto &sender : json::get<"senders"_>(filter))
			if(at<"sender"_>(event) == unquote(sender))
				return true;

		return false;
	}

	return true;
}

//
// filter
//

ircd::m::filter::filter(const user &user,
                        const string_view &filter_id,
                        const mutable_buffer &buf)
{
	get(user, filter_id, [this, &buf]
	(const json::object &filter)
	{
		const size_t len
		{
			copy(buf, string_view{filter})
		};

		new (this) m::filter
		{
			json::object
			{
				data(buf), len
			}
		};
	});
}

ircd::string_view
ircd::m::filter::set(const mutable_buffer &idbuf,
                     const user &user,
                     const json::object &filter)
{
	const auto user_room_id
	{
		user.room_id()
	};

	const m::room room
	{
		user_room_id
	};

	const sha256::buf hash
	{
		sha256{filter}
	};

	const string_view filter_id
	{
		b64encode_unpadded(idbuf, hash)
	};

	send(room, user.user_id, "ircd.filter", filter_id, filter);
	return filter_id;
}

void
ircd::m::filter::get(const user &user,
                     const string_view &filter_id,
                     const closure &closure)
{
	if(!get(std::nothrow, user, filter_id, closure))
		throw m::NOT_FOUND
		{
			"Filter not found"
		};
}

bool
ircd::m::filter::get(std::nothrow_t,
                     const user &user,
                     const string_view &filter_id,
                     const closure &closure)
{
	const auto user_room_id
	{
		user.room_id()
	};

	const m::room room
	{
		user_room_id
	};

	return room.get(std::nothrow, "ircd.filter", filter_id, [&closure]
	(const m::event &event)
	{
		const json::object &content
		{
			at<"content"_>(event)
		};

		closure(content);
	});
}

//
// room_filter
//

ircd::m::room_filter::room_filter(const mutable_buffer &buf,
                                  const json::members &members)
:super_type::tuple
{
	json::stringify(mutable_buffer{buf}, members)
}
{
}

//
// room_event_filter
//

ircd::m::room_event_filter::room_event_filter(const mutable_buffer &buf,
                                              const json::members &members)
:super_type::tuple
{
	json::stringify(mutable_buffer{buf}, members)
}
{
}

//
// event_filter
//

ircd::m::event_filter::event_filter(const mutable_buffer &buf,
                                    const json::members &members)
:super_type::tuple
{
	json::stringify(mutable_buffer{buf}, members)
}
{
}
